### PR TITLE
RFC: JSON Patch support for Ember Data

### PR DESCRIPTION
**Pull request by pixelhandler**
[Rendered][json patch support rfc]

@machty suggested that I add the proposal I posted on the discuss forum to this repo, see: http://discuss.emberjs.com/t/json-patch-support-for-ember-data/6078 ( [json patch support rfc] )

"Extend Ember Data Model prototypes to have individual 'updatable' properties. Model instances can receive patch operations in response to updating or deleting records (PUT or DELETE requests) by pushing **partial** changes into the store (DS.Store#update). Create mixins for Model, Adapter & Serializer prototypes which support using a JSON Patch [rfc6902] format in payloads sent as responses in order to apply partial updates."

[rfc6902]: http://tools.ietf.org/html/rfc6902
[json patch support rfc]: https://github.com/pixelhandler/rfcs/blob/json-patch-support-for-ember-data/active/0000-json-patch-support-for-ember-data.md

### CODE REVIEW

**Comment by mhretab** [comment]
+1

**Comment by dustinfarris** [comment]
+1

**Comment by jeffreylouden** [comment]
+1

**Comment by goto-bus-stop** [comment]
:+1:

**Comment by timini** [comment]
Want :)

**Comment by egaba88** [comment]
+1

**Comment by joebartels** [comment]
:+1:

File: active/0000-json-patch-support-for-ember-data.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+]
+```
+
+[JSON Patch]: http://jsonpatch.com
+[rfc6902]: http://tools.ietf.org/html/rfc6902
+
+## Technical Considerations
+
+* Mixin for (Ember Data) DS.Adapter/DS.Serializer for PATCH requests
+* Mixin for DS.Model to support changes that fire PATCH requests
```
Comment (kurko) [inline active/0000-json-patch-support-for-ember-data.md]: IMO, models shouldn't know about data exchange mechanisms with the server. We'd entangle models with adapters with servers, which would lead to highly coupled code.

In my opinion, we should add a new method in the `store` such as `bulkSave` that would trigger yet another method in the adapter, `patch`. Adapters without this method mean they don't support this and regular `createRecord` / `updateRecord` would be used.

The adapter's `patch` method would receive a special object called `DirtyRecords` (or along those lines), which would yield an array of all records that are pending save in the store. Adapter would craft a JSON Patch request based on that object.

**Comment by jeremytm** [comment]
+1

**Comment by archit** [comment]
+1

**Comment by ryanlitalien** [comment]
+1

**Comment by kalebdf** [comment]
+1

**Comment by pybt** [comment]
+1

**Comment by coen-hyde** [comment]
Yes please!

**Comment by kurko** [comment]
All of these +1 don't add anything. Your interaction will be much more effective if you, for instance, reply to my comment above about coupling models with servers.

**Comment by tikotzky** [comment]
@kurko this chrome extension is a godsend... [Github +1s](https://chrome.google.com/webstore/detail/github-%201s/lddfkkebajnpiicnpfamebilmhamkeme)
_though you still do get the email notifications..._

**Comment by jeremytm** [comment]
@kurko I'm not sure I understand why. Also, there is a button 'Add a line note' to your comment. Is this what you mean? Is it better to add '+1' there instead?

**Comment by kurko** [comment]
@jeremytm I'm suggesting you to elaborate on why you agree or disagree with https://github.com/emberjs/rfcs/pull/5#discussion_r25258635, which looks like a blocker IMO.

**Comment by jeremytm** [comment]
I think like myself, many us here just want to show support for adding PATCH capabilities into Ember Data - assuming that the brilliant minds behind it will know how to work through any obstacles and blocks - and land on the best implementation.

Any developer knows the features that are requested most are the ones that receive attention, so that's the goal here from these +1's I think.

What's the next step? (I do agree with your statement btw, that models should only know what they need to know and nothing more). Does someone start a new thread with an evolved implementation suggestion? Then we all jump over there with our +1's?

File: active/0000-json-patch-support-for-ember-data.md (hunk 1 above)
Comment (pixelhandler) [inline active/0000-json-patch-support-for-ember-data.md]: @kurko I understand your concern. My intention is that a model can trigger an event on the store service such as `attributeChanged` then the service/adapter can do what it wants, e.g. perhaps call a `patch` method. This is an example of one way that I've implemented this behavior to patch a request as close to real time as possible - https://github.com/pixelhandler/ember-jsonapi-resources/blob/master/README.md#resource-services and here - https://github.com/pixelhandler/ember-jsonapi-resources/wiki/Services#collaborators

I really enjoyed this behavior when I tried out ember-orbit with json-patch. So I've borrowed that behavior for my own persistence needs as well.

**Comment by pixelhandler** [comment]
@jeremytm this is the **5th** PR on this RFC repo (this weekend I created the **69th** PR), and has not had any attention/response from the Ember Data team. Not so much as a "no that's not a good idea" or any response with an alternative solution. 

I expect that I will just continue to develop my own custom solutions for data persistence.  Here is my story and path forward for my data persistence needs: http://pixelhandler.com/posts/my-battle-with-data-persistence-in-ember-apps and the repo I am now using for JSON API - https://github.com/pixelhandler/ember-jsonapi-resources I may implement JSON PATCH on that project once my backend is ready for it. I ran with JSON Patch on personal project for about a year and really enjoyed it. For me, the JSON API spec is really the main driver for a solid client-server solution for Ember apps. Ember Data is also using the format for JSON API as well. The Ember Orbit project did have JSON Patch support but I'm not sure if that will continue. 

About a year ago, I spoke with one of the Ember Data team members and doubts about JSON Patch were raised concerning enforcement of an order of patch operations. (I think that the Orbit.js project did address that concern internally.) In an application with lots of read/write activity timestamps may be necessary to enforce the order of operations when processing a batch of operations. I really like the idea of using the test operation to validate that each current operation is valid on the server side.

**Comment by kurko** [comment]
@wecc @igorT any chance this could get some eyes? I think this moves in the right direction.

**Comment by dgeb** [comment]
@pixelhandler - let me address the questions you raised about [Orbit](https://github.com/orbitjs/orbit.js) and [Ember-Orbit](https://github.com/orbitjs/ember-orbit).

Orbit internally uses JSON Patch operations to transform sources and emit notifications that sources have transformed. This is still the case and I don't expect this to change.

Furthermore, Orbit has a `Source` that communicates via JSON Patch operations with a server: the `JSONAPIPatchSource`. This source communicates using the experimental [JSON Patch Extension to the JSON API spec](http://jsonapi.org/extensions/jsonpatch/). I have placed development of the `JSONAPIPatchSource` on hold temporarily until we settle the extension support details in JSON API. We have targeted this feature for JSON API 1.1, which will be released in beta on September 30th.

So to recap, Orbit is fully committed to continuing to support JSON Patch operations internally and will also support the final form of the JSON Patch extension that lands in JSON API. 

Now that we've finalized JSON API 1.0, I'm devoting almost all my OSS time to stabilizing and refining Orbit and Ember-Orbit. It would be a massive duplication of effort to enable the same internal and external communication mechanisms in Ember Data. I'm open for discussions with the Ember Data team regarding collaboration options between our projects.

**Comment by SeyZ** [comment]
:+1:  +1

**Comment by jeremytm** [comment]
JSON API 1.0 requires PATCH requests for updates, and it looks like in Ember data 1.13, using the JSONAPIAdapter, it's sending out PATCH requests for all updates by default - just for anyone else who ends up here, and has the ability to use JSON API 1.0 for their backend.

**Comment by pixelhandler** [comment]
Nice discussion, thanks everyone. Closing this, it doesn't need to be a thing in Ember Data, cheers.
